File: web/src/types/ruleWarning.ts
import { z } from 'zod'

/** A soft validation finding (openapi `RuleWarning`): guidance, never a block. `field` is the
 *  offending field id, or `null` for a sheet-wide warning — `.nullish()` because the BFF serializes the
 *  Kotlin nullable as explicit `null`, which `.optional()` would reject. */
export const ruleWarningSchema = z.object({
  code: z.string(),
  field: z.string().nullish(),
  message: z.string(),
})

export type RuleWarning = z.infer<typeof ruleWarningSchema>

File: api/src/main/kotlin/no/rauboti/tome/rulesets/domain/RuleWarning.kt
package no.rauboti.tome.rulesets.domain

/**
 * A soft validation finding: guidance, never a hard block — the DM can always override. [field] is
 * the offending field id, or null for a sheet-wide warning. Serialized to the openapi `RuleWarning`
 * schema.
 */
data class RuleWarning(
    val code: String,
    val message: String,
    val field: String? = null,
)
